### ISSUE TITLE
node server crashes on malformed URI request

### ISSUE THREAD

**Issue by kenkunz** [tradingstrategy-ai/frontend#102]
Sometime prior to 2022-05-23 16:15 UTC our website went down.

@hieuh25 investigated this and found the following error had occurred on the node server:
```
file:///root/frontend/build/server/index.js:1764
  let decoded = decodeURI(url.pathname);

URIError: URI malformed 
    at decodeURI (<anonymous>)  
    at respond (file:///root/frontend/build/server/index.js:1764:17)
    at Server.respond (file:///root/frontend/build/server/index.js:2022:12)
    at Array.ssr (file:///root/frontend/build/handler-edaaddc2.js:841:16)
```

Further investigation of Caddy logs found the proximate cause to be following malformed GET request:
```
GET /%CD%EA%D5%FB.rar
```

This is easily reproducible by running the `node` server locally and requesting the above path.

**NOTE: please do not try that URL in production** until this issue is resolved.

**Comment by kenkunz** [comment]
I did some further investigation.

The URL string above is not a valid encoded value (as reported by the `URIError` exception).  The SvelteKit `node-adapter` doesn't have any built-in error handling for this scenario. It _does_ export a handler function that we could [ wrap with our own custom server](https://github.com/sveltejs/kit/tree/master/packages/adapter-node#custom-server).

I tested this with the following code:

#### `index.js`
```javascript
import { createServer } from 'http';
import { handler } from './build/handler.js';

const port = process.env.FRONTEND_PORT || 3000;

const server = createServer((req, res) => {
  try {
    decodeURI(req.url);
  } catch (e) {
    console.log(`URIError: can't decode ${req.url}; re-encoding and passing on`);
    req.url = encodeURI(req.url);
  } finally {
    handler(req, res);
  }
});

server.listen(port, () => {
  console.log(`Listening on port: ${port}`);
});
```

For an actual solution, it might be preferable to redirect to a specific error page and pass along the encoded (bogus) URL value as a param so it can be decoded and displayed in the error message.

That said… the more I thought about this, the more it feels like something that should be handled upstream – either by Cloudflare CDN or by the Caddy reverse proxy.

@miohtama, @hieuh25, @servatj – any opinion on the best place to solve this?

**Comment by kenkunz** [comment]
Another safeguard we should consider – we've had other exceptions cause a fatal error of the `node` service (e.g., `fetch` DNS errors). We can't anticipate ahead-of-time all the unhandled exceptions that cause `node-adapter` to explode. I think it would be a good idea to wrap the `node` command in a simple supervisor to restart on exit, such as:

```bash
while true; do
  node build
  sleep 1
done
```

We could add some error output to this so it's easier to track down in the logs.

I know there's a better solution to this once we're running on Docker… but in the meantime, this seems like an easy safeguard.

Thoughts?

**Comment by kenkunz** [comment]
@miohtama we discussed this in the Daily today… @hieuh25 did not have a strong opinion and deferred to your recommendation.

My 2¢ – I'd prefer to address this upstream (e.g., Caddy), but am happy to implement a wrapper service like the example above if you prefer that.

We also agreed that it couldn't hurt to open a SvelteKit issue to see if they'd consider addressing this w/in the `node-adapter`. That said, unless we submit the PR fix ourselves, I don't think we want to wait for them to fix it (they'll be focussed on getting to `1.0`).
